cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -385,6 +385,23 @@ Result<ScanTaskIterator> ParquetFileFormat::ScanFile(
   return MakeVectorIterator(std::move(tasks));
 }
 
+util::optional<Future<int64_t>> ParquetFileFormat::CountRows(
+    const std::shared_ptr<FileFragment>& file, Expression predicate,
+    std::shared_ptr<ScanOptions> options) {
+  auto parquet_file = internal::checked_pointer_cast<ParquetFileFragment>(file);
+  if (FieldsInExpression(predicate).size() > 0) {
```

Comment:
Would it be worthwhile to try pushing the predicate down to row groups here? If every row group happens to pass or fail the filter whole, we could recover the metadata-only fast path (bailing if any row group requires non trivial filtering)

Reply:
It should be doable. However, it'll require rethinking the API a bit: currently, it assumes we know up front whether we can use the fast path. But to push down the predicate, we may need to load the footer, so it'll have to go from optional<Future<int>> to Future<optional<int>>. I think that shouldn't penalize the slow path really though since either the Future can be constructed right away, or it'll do work that we need to do later in the scan anyways (namely load the footer).

Reply:
Ah, it turns out it doesn't work because right now, we can only definitively reject a row group, but can't decide whether a row group should definitively be included in the row count, because `SimplifyWithGuarantee((i64 == 1), (i64 <= 1) & (i64 >= 1))` simplifies to `i64 == 1`, not to `true`.

Reply:
It works for some predicates - so now it's implemented. It'll naturally get better as the expression machinery improves.